Bbcode

ngilain opened the pull request.

nicolasThal commented on en/developer_guide/bbcode.rst:
```diff
@@ -0,0 +1,136 @@
+``open_orchestra_bbcode.code_definitions`` contains an array of declared simple tags. Here is for
+instance the declaration of the bold tag:
+
+::
```
.. code-block :: yaml ??

nicolasThal commented on en/developer_guide/bbcode.rst:
```diff
@@ -0,0 +1,136 @@
+
+So to create a bold tag definition object, you can write:
+
+::
```
.. code-block:: php

nicolasThal replied:
see https://raw.githubusercontent.com/open-orchestra/open-orchestra-docs/master/en/developer_guide/custom_fixtures_load.rst for usage

ngilain commented:
@nicolasThal : update done

alavieille committed:
change account sensio insight

fix insight worklow function bundle

Create the BBCode doc

fix rst format

Fix lists in bbcode doc

Code review

Code review

Code Review

Fix bullet point format

Fix bullet point format

Code review

Code review

code review

Code review

nicolasThal commented on en/developer_guide/bbcode.rst:
```diff
@@ -0,0 +1,136 @@
+To generate a new simple definition with this factory, call the method ``create`` with the following
+parameters:
+
+- $tag: the BBCode tag, ie ``b`` for the tag [b]
```
shouldn't youput the $tag between `` ?

ngilain commented:
@nicolasThal : update done

alavieille commented:
think to add it in summary

alavieille commented on en/developer_guide/bbcode.rst:
```diff
@@ -0,0 +1,136 @@
+
+4. When the user saves the form, TinyMce replaces via javascript the html tags by their BBCode
+representation.
+
```
May be a little schema to be more explicit

alavieille commented on en/developer_guide/bbcode.rst:
```diff
@@ -0,0 +1,136 @@
+        tag: b
+        html: <strong>{param}</strong>
+
+- The key 'bold' is the name of the definition used internally by the parser.
```
The key `bold`

alavieille commented on en/developer_guide/bbcode.rst:
```diff
@@ -0,0 +1,136 @@
+
+Create a simple BBCode tag via service
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+BBCode definition can also be declared by services. A compiler pass has been added to the
```
service

ngilain replied:
In fact I wanted to say BBCode definitionS can also be declared by services

alavieille commented on en/developer_guide/bbcode.rst:
```diff
@@ -0,0 +1,136 @@
+- The ``html`` key is the html pattern of the code. The string ``{param}`` will be replaced by the
+content of the BBcode tag.
+
+- The ``tag`` key is the BBCode pattern of the tag. In this example the tag is [b]{param}[/b].
```
`[b]{param}[/b]` add ``

alavieille commented on en/developer_guide/bbcode.rst:
```diff
@@ -0,0 +1,136 @@
+
+- The ``tag`` key is the BBCode pattern of the tag. In this example the tag is [b]{param}[/b].
+
+Simple tags can also include some options. For instance the tag [h] which matches a <h*> html tag
```
same

ngilain replied:
no it's really some options ^^

alavieille commented on en/developer_guide/bbcode.rst:
```diff
@@ -0,0 +1,136 @@
+- The ``tag`` key is the BBCode pattern of the tag. In this example the tag is [b]{param}[/b].
+
+Simple tags can also include some options. For instance the tag [h] which matches a <h*> html tag
+requires an option to indicate the level of the header. For instance [h=1]Some text[/h] will
```
same

ngilain replied:
Here too

alavieille commented on en/developer_guide/bbcode.rst:
```diff
@@ -0,0 +1,136 @@
+mandatory, you should declare two versions of the tag: one with option and one without. That way
+the parser will use one or the other according to the presence of an option. The ``html`` key uses
+both {param} and {option} string to indicate where to place those informations.
+The resulting BBCode will be [h={option}]{param}[/h]
```
use ``

alavieille commented on en/developer_guide/bbcode.rst:
```diff
@@ -0,0 +1,136 @@
+Instead a specific PHP code needs to be executed. This can be done in a definition implementing
+``OpenOrchestra\BBcodeBundle\Definition\BBcodeDefinitionInterface``
+All you have to do is to write the method getHtml, which render the html code of the BBcodeElementNode given
+as a parameter.
```
may be an example ?

ngilain commented:
@itkg-alavieille , @nicolasThal : doc updated

nicolasThal commented on en/developer_guide/bbcode.rst:
```diff
@@ -0,0 +1,153 @@
+
+It's important to note that by default, not all tags are enabled in BBCode format, only those you
+want to declare. For instance, putting a text in bold in the rich text editor does not produce by
+default a [b] tag. Out of the box, there is only one BBCode: ``media``. But you can add your own.
```
one BBCode tag

nicolasThal replied:
We allow you to add your own tag

nicolasThal commented on en/developer_guide/bbcode.rst:
```diff
@@ -0,0 +1,153 @@
+``[media=format]MediId[/media]`` requires to fetch the media in the database, to detect its type and
+then to to return the corresponding html tag.
+
+Add a new BBCode to the parser requires to add its definition. There are two ways of adding a
```
To add a new BBCode tag to the parser, you need to add its definition

nicolasThal replied:
There are two ways to add a

nicolasThal commented on en/developer_guide/bbcode.rst:
```diff
@@ -0,0 +1,153 @@
+- The ``html`` key is the html pattern of the code. The string ``{param}`` will be replaced by the
+content of the BBcode tag.
+
+- The ``tag`` key is the BBCode pattern of the tag. In this example the tag is ``[b]{param}[/b]``.
```
I think that the yaml code is not corresponding

ngilain replied:
The yaml is correct but I have to present keys not in the config order to easily introduce the string {param}. Presenting the keys in the config order would lead to a complex description of the string {param}

However I rewrite the sentence about "tag" which may be ambiguous. Let me know if it is ok for you when I have push it.

nicolasThal commented on en/developer_guide/bbcode.rst:
```diff
@@ -0,0 +1,153 @@
+    }
+
+Using a service to define your tag allows you to access the Symfony container and to use services
+to do a complex transformation.
```
The service should not be tagged in the symfony dependancy injection ?

ngilain replied:
The definition doen't have to be a tagged service, the definitions collection must. With your comment, I understand that this point is not clear, so I rewrite the service part to be clearer.

nicolasThal commented on en/summary.rst:
```diff
@@ -63,6 +63,7 @@ Developer guide
 * `Custom error pages`_
 * `Robots.txt`_
 * `Sitemap.xml`_
+* `Add a custom BBCode`
```
add an _ at the end of the line

ngilain commented:
@nicolasThal : doc updated

veolys-nanne commented on en/developer_guide/bbcode.rst:
```diff
@@ -0,0 +1,172 @@
+1. Each rich text field goes into a data transformer to interpret the BBCode tags. This data
+   transformer calls the Open Orchestra BBCode parser to detect and transform valid BBCode tags into
+   html tags.
+2. The resulting html is given to the TinyMce editor who treat it with its standard logic.
```
'which treats' in place of who treat

veolys-nanne commented on en/developer_guide/bbcode.rst:
```diff
@@ -0,0 +1,172 @@
+3. When the user clicks in the toolbar to inject a BBCode tag (for instance by selecting a media),
+   TinyMce inject the html representation of the BBCode, to be able to visually render the tag. For
+   instance in the media case, you will see it in the editor.
+4. When the user saves the form, TinyMce replaces via javascript the html tags by their BBCode
```
as tag is plural their BBCode representation should be too.

veolys-nanne commented on en/developer_guide/bbcode.rst:
```diff
@@ -0,0 +1,172 @@
+``<b>Some Text</b>``
+
+- Complex tags require more than a simple regexp conversion. For instance the tag
+``[media=format]MediId[/media]`` requires to fetch the media in the database, to detect its type and
```
`[media=format]MediaId[/media]` in place of `[media=format]MediId[/media]`

veolys-nanne commented on en/developer_guide/bbcode.rst:
```diff
@@ -0,0 +1,172 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+To create a simple BBCode definition, its possible to use a trivial configuration. The parameter
+``open_orchestra_bbcode.code_definitions`` contains an array of declared simple tags. Here is for
+instance the declaration of the bold tag:
```
a lot of "for instance" (by example, in this case ....)

veolys-nanne commented on en/developer_guide/bbcode.rst:
```diff
@@ -0,0 +1,172 @@
+                use_option: true
+
+To declare that a tag uses option, you must add the key ``parameters.use_option`` and set it to true.
+Not that when declaring a tag with option, that option becomes mandatory. So if the option is not
```
If its not 'Note', i don't understand the meaning.

veolys-nanne commented on en/developer_guide/bbcode.rst:
```diff
@@ -0,0 +1,172 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+BBCode definitions can also be declared by services. A compiler pass has been added to the
+BBCodeBundle to collect all services tagged with ``open_orchestra_bbcode.code_definitions``. Such
+a service includes a BBCode definition collection and serve it via the method ``getDefinitions``.
```
serves in place of serve

veolys-nanne commented on en/developer_guide/bbcode.rst:
```diff
@@ -0,0 +1,172 @@
+As mentionned before, to let Open Orchestra know your new definition, you have to add it to a
+definition collection class implementing
+``OpenOrchestra\BBcodeBundle\Definition\BBcodeDefinitionCollectionInterface``. Then make that
+collection a Symfony service tagged with ``open_orchestra_bbcode.code_definitions``.
```
Maybe could you give an example.

ngilain replied:
I don't know what to show: it's straight forward as the interface only have one explicit method: getDefinitions. I think it may unnecessarily burden the explanation :s

veolys-nanne commented on en/developer_guide/bbcode.rst:
```diff
@@ -0,0 +1,172 @@
+searching in its parameter for a valid youtube id, you have to create your own definition class.
+This definition class must implement
+``OpenOrchestra\BBcodeBundle\Definition\BBcodeDefinitionInterface``. The getHtml method receives a
+BBcodeElementNode representing the BBCode tag and return the matching html version of the tag. So
```
returns in place of return

ngilain commented:
@itkg-alavieille , @itkg-nanne , @nicolasThal : doc updated

nicolasThal commented on en/developer_guide/bbcode.rst:
```diff
@@ -0,0 +1,172 @@
+to add your own tags.
+
+In this documentation you will see how the BBCode feature fits in the main processes and how to create
+a new BBCode in your bundle.
```
a new BBCode tag

nicolasThal replied:
a new BBCode tag ?

nicolasThal replied:
still no ?

nicolasThal commented on en/developer_guide/bbcode.rst:
```diff
@@ -0,0 +1,172 @@
+
+How to create a new BBCode and add it to the parser
+---------------------------------------------------
+Open Orchestra provides a simple mechanism to add your own custom BBCode tags. In fact there is two
```
Open Orchestra provides simple ways to add your own custom BBCode tags, depending on the tag nature :

nicolasThal commented on en/developer_guide/bbcode.rst:
```diff
@@ -0,0 +1,172 @@
+and then to to return the corresponding html tag.
+
+To add a new BBCode to the parser, you need to create its definition. There are two ways to add a
+definition: via configuration, or via a tagged service.
```
Is the configuration way only for simple tag ?

ngilain commented:
@itkg-alavieille , @itkg-nanne , @nicolasThal : doc updated

veolys-nanne commented:
:+1:

nicolasThal commented on en/developer_guide/bbcode.rst:
```diff
@@ -0,0 +1,173 @@
+``[media=format]MediaId[/media]`` requires to fetch the media in the database, to detect its type
+and then to to return the corresponding html tag.
+
+To add a new BBCode to the parser, you need to create its definition. There are two ways of adding
```
To add a new BBCode tag to the parser

nicolasThal commented on en/developer_guide/bbcode.rst:
```diff
@@ -0,0 +1,173 @@
+
+To add a new BBCode to the parser, you need to create its definition. There are two ways of adding
+a definition. To create a simple tag you can quickly set a configuration. But if you prefer, or if
+your tag requires a php analysis, you can use a tagged service.
```
you can use a symfony service (I don't think that you have to already say that the service should be tagged, it will be said later)

nicolasThal commented on en/developer_guide/bbcode.rst:
```diff
@@ -0,0 +1,173 @@
+
+Create a BBCode tag via a service
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+BBCode definitions can also be declared by services. A compiler pass has been added to the
```
I don't think that talking about `CompilerPass` here is usefull to the documentation.

You could say : 
`The service receiving the symfony tag ... are going to be added to the parser`

ngilain replied:
To be exact, it's not the service who's added to the parser, but the definitions it presents. But I see what you mean, so I adjust in that way.

nicolasThal commented on en/developer_guide/bbcode.rst:
```diff
@@ -0,0 +1,173 @@
+The definition collection must be composed of formatted elements implementing
+``OpenOrchestra\BBcodeBundle\Definition\BBcodeDefinitionInterface``. If your definition is a simple
+regex replacement, as the one seen in the previous section, you can use an Open Orchestra definition
+factory. But to create a BBCode with a more complex logic, you will have to create your own
```
BBCode tag

ngilain commented:
@nicolasThal : update done

nicolasThal commented:
:+1: after a little squash
